velox/common/base/tests/RawVectorTest.cpp

```diff
@@ -110,3 +111,145 @@ TEST(RawVectorTest, toStdVector) {
     ;
   }
 }
+
+struct State {
+  const int32_t* data;
+  int32_t* result;
+  int32_t key;
+  int32_t lo;
+  int32_t hi;
+  int32_t guess;
+  bool done{false};
+
+  State() = default;
+
+  State(const raw_vector<int32_t>& vec, int32_t key, int32_t* result)
+      : data(vec.data()),
+        result(result),
+        key(key),
+        lo(0),
+        hi(vec.size()),
+        done(false) {}
+
+  void lookup() {
+    guess = (lo + hi) / 2;
+    __builtin_prefetch(data + guess);
+  }
+
+  bool decide() {
+    if (done) {
+      return false;
+    }
+    auto localGuess = guess;
+    int32_t value = data[guess];
+    for (;;) {
+      if (value < key) {
+        lo = localGuess + 1;
+      } else if (value > key) {
+        hi = localGuess;
+      } else {
+        *result = localGuess;
+        done = true;
+        return true;
+      }
+      auto dist = hi - lo;
+      if (dist > 16) {
+        guess = (lo + hi) / 2;
+        __builtin_prefetch(data + guess);
+        return false;
+      }
+      if (!dist) {
+        *result = lo;
+        done = true;
+        return true;
+      }
+      localGuess = (lo + hi) / 2;
+      value = data[localGuess];
+    }
+  }
+};
+
+template <int32_t stride>
+void searchTest(
+    const std::vector<raw_vector<int32_t>>& vectors,
+    const std::vector<int32_t>& indices,
+    const std::vector<int32_t>& keys,
+    std::vector<int32_t>& positions) {
+  SelectivityInfo info;
+  {
+    SelectivityTimer t(info, indices.size());
+    if (stride == 1) {
+      for (auto iv = 0; iv < indices.size(); ++iv) {
+        auto nth = indices[iv];
+        auto& vec = vectors[nth];
+        auto target = keys[iv];
+        auto data = vec.data();
+        int lo = 0, hi = vec.size();
+        while (lo < hi) {
+          int guess = (lo + hi) / 2;
+          if (data[guess] < target) {
+            lo = guess + 1;
+          } else if (data[guess] == target) {
+            positions[iv] = guess;
+            goto found;
+          } else {
+            hi = guess;
+          }
+        }
+        positions[iv] = lo;
+      found:;
+      }
+    } else {
+      std::vector<State> states(stride);
```

Comment:
We should also try Eytzinger layout if we really want to optimize binary search.  That way we don't need to change the interface to take multiple values at same time while improve cache performance.